### ISSUE TITLE
Keyboard shortcuts/accelerators for dialog buttons

### ISSUE THREAD

**Issue by damieng** [electron/electron#7517]
## Problem

Windows and Linux allow keyboard accelerators (one-key shortcuts) in dialogs which would be useful to be able to access via the Dialog module in Electron in a safe way.

Today the underlying API on Windows (and I suspect Linux) take & as the signal that the following letter in the button is the accelerator and Atom passes that through e.g.

`dialog.showMessageBox({message: 'Your changes will be lost unless you save them.', buttons: [ '&Don\'t Save', '&Save' ]})`

Renders as;

![image](https://cloud.githubusercontent.com/assets/118951/19169473/84d04d66-8bc9-11e6-9f3c-2ca843de8546.png)

macOS doesn't support that syntax and will display the ampersand in-place.  This means developers using Electron must wrap all their calls to Dialog with appropriate logic to either add or strip the & before passing through.

**Comment by damieng** [comment]
# Proposed solution 1 - Strip & on macOS

In an ideal world we could just pass `&Don't Save` and on macOS `&` followed by any character other than another `&` would be stripped with Windows and Linux passing through.  

This would mean existing Electron apps upgrading would find & missing from their button labels on macOS. I'm not sure this is really a problem however as those buttons are already today messed up on Windows and Linux.

**Comment by damieng** [comment]
## Proposed solution 2 - Accept objects as well as strings in the buttons array

This would allow 100% compatibility with the current (broken?) behavior. The syntax could look something like this;

```dialog.showMessageBox({message: 'Your changes will be lost unless you save them.', buttons: [ { labelWithAccelerator: '&Don\'t Save', &'Save' } ]})```

**Comment by damieng** [comment]
I removed the previous option about having an object with an 'accelerator' key property as it doesn't let you choose which instance of that letter in the label is underlined which is a pain.  e.g.  `Always &Leave` could not be represented with ```{ label: 'Always Leave', accelerator: 'L' }``` would highlight the `l` in `Always`

**Comment by ccoenen** [comment]
I'm curious: what happens, if I labelled a button with an actual, readable ampersand? Let's say I create a button that is supposed to say "wash&dry" (for whatever reason, this is just an example!) Do I have to escape that Ampersand in "original" Win32 API?

**Comment by enlight** [comment]
> Do I have to escape that Ampersand in "original" Win32 API?

Yes. If you don't want the & to be interpreted as an accelerator you have to escape it with another &, so you'd need to write "wash&&dry".

**Comment by felixrieseberg** [comment]
Given that ☝️ is the case, I think just stripping a non-escaped ampersand from macOS is the easiest and most straightforward solution.

**Comment by jtojnar** [comment]
This is a ‘must have’ for controlling applications using keyboard.

Also note that on Linux (at least with GTK), **underscore** is used for accelerators.

```javascript
dialog.showMessageBox(getCurrentWindow(), {
	'message': 'Accelerations demo.',
	'buttons': ['O_K', 'G_o', '&Moo']
});
```

![accel-demo](https://cloud.githubusercontent.com/assets/705123/19484319/7b9c0072-9557-11e6-8591-ad30a32ab142.png)
